### PR TITLE
CDT-1252 added support for firefox and remote firefox browser

### PR DESCRIPTION
**Pull request by raunakbagri**

### CODE REVIEW

**Comment by jchang691** [comment]
Message me when you merge it in. I'll do a deploy out. May take a couple of hours since it's an open source project.

File: src/main/java/com/lithium/mineraloil/selenium/browsers/FireFoxBrowser.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.awaitility.Awaitility.await;
+
+@Slf4j
+public class FireFoxBrowser implements Browser {
```
Comment (zed-code) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/FireFoxBrowser.java]: rename to proper naming standard

File: src/main/java/com/lithium/mineraloil/selenium/browsers/FireFoxBrowser.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            await().atMost(1, MINUTES)
+                    .pollInterval(5, SECONDS)
+                    .until(() -> driverInitializer.getDriverInThread() != null);
+        } catch (ConditionTimeoutException e) {
+            throw new DriverNotFoundException("Was unable to get a Remote Driver!!!");
+        }
+        return driverInitializer.getWebDriver();
+    }
+
+    class DriverInitializer {
```
Comment (zed-code) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/FireFoxBrowser.java]: create an interface. and separate class as well.  Then extend it as per your need

File: src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.firefox.FirefoxOptions;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+import java.net.URL;
+import java.util.concurrent.Callable;
+
+@Slf4j
+public class RemoteFireFoxBrowser extends RemoteBrowser {
```
Comment (zed-code) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java]: same naming standard

File: src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    private final FirefoxOptions firefoxOptions;
+    private final int remoteFireFoxPort;
+    private final String remoteWebdriverAddress;
+
+    public RemoteFireFoxBrowser(DriverConfiguration driverConfiguration) {
```
Comment (zed-code) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java]: naming

File: src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        firefoxOptions = driverConfiguration.getFirefoxOptions();
+        remoteFireFoxPort = driverConfiguration.getRemotePort() != 0 ? driverConfiguration.getRemotePort() : 4444;
```
Comment (zed-code) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java]: line alignment.

File: src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    void logCapabilities() {
+        log.info(String.format("Desired Capabilities: %s", firefoxOptions));
+    }
+
+    @Override
+    Callable<WebDriver> getDriverThreadCallableInstance() {
+        return new GridDriverThread(serverAddress, firefoxOptions);
+    }
+
+    private class GridDriverThread implements Callable<WebDriver> {
```
Comment (zed-code) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/RemoteFireFoxBrowser.java]: separate file class, please

**Comment by raunakbagri** [comment]
@zed-code please review

File: src/main/java/com/lithium/mineraloil/selenium/browsers/ChromeBrowser.java [hunk 7]
```diff
@@ -39,36 +32,11 @@ public WebDriver getDriver() {
         DriverInitializer driverInitializer = new DriverInitializer();
         try {
             await().atMost(1, MINUTES)
-                   .pollInterval(5, SECONDS)
-                   .until(() -> driverInitializer.getDriverInThread() != null);
+                    .pollInterval(5, SECONDS)
```
Comment (jchang691) [inline src/main/java/com/lithium/mineraloil/selenium/browsers/ChromeBrowser.java]: Line alignment change not needed here.

Comment (raunakbagri) [reply src/main/java/com/lithium/mineraloil/selenium/browsers/ChromeBrowser.java]: done

**Pull request merged**